compiler/semcall.nim

```diff
@@ -323,10 +323,16 @@ proc resolveOverloads(c: PContext, n, orig: PNode,
       pickBest(callOp)
 
     if overloadsState == csEmpty and result.state == csEmpty:
+      let ident = considerQuotedIdent(c, f, n).s
+      var msg = ""
       if nfDotField in n.flags and nfExplicitCall notin n.flags:
-        localError(c.config, n.info, errUndeclaredField % considerQuotedIdent(c, f, n).s)
+        msg = errUndeclaredField % ident
       else:
-        localError(c.config, n.info, errUndeclaredRoutine % considerQuotedIdent(c, f, n).s)
+        msg = errUndeclaredRoutine % ident
+      let sym = qualifiedLookUp(c, f, {})
+      if sym != nil:
+        msg &= "; found '$1' of kind '$2' used in wrong context" % [getSymRepr(c.config, sym), $sym.kind]
```

Comment:
The error message is quite verbose and leaks implementation details such as the SK flags.
I'd just modify it so that it says "undeclared <routine/template/macro/...>", basically stringify the filter set in a smart way.

This is also super helpful for unknown pragma names, seeing the " undeclared routine" error always left me puzzled.

Reply:
I made the msg a bit less verbose, and special cased `kind skIterator` to `kind iterator` to avoid leaking implementation detail in that case; 

I'd like to keep the relevant information as to where the symbol was found `eg: undeclared_routime.myiter()[declared in undeclared_routime.nim(5, 9)]` and its kind, as these help for understanding what went wrong;

if that's not ok for some reason, please write the exact msg you'd want to see in place of
"attempting to call undeclared routine: 'myiter', found 'undeclared_routime.myiter()[declared in undeclared_routime.nim(5, 9)]' of kind 'iterator'"

Reply:
Well, you're telling the user that `x` is undeclared and where `x` is declared, the user may not appreciate the irony :)

The same error shows up when you apply a non existent pragma name to something (custom pragmas are templates marked with `{.pragma.}`).

I'd just make it say `attempting to call undeclared iterator/template/routine 'myiter'`.

Including some extra context may also be wrong since you may have a symbol choice and it doesn't make sense to print just one candidate.

Reply:
> Well, you're telling the user that x is undeclared and where x is declared, the user may not appreciate the irony :)

fixed

> The same error shows up when you apply a non existent pragma name to something (custom pragmas are templates marked with {.pragma.}).

add test for this

> Including some extra context may also be wrong since you may have a symbol choice and it doesn't make sense to print just one candidate.

fixed: I now show all candidates in case we expected a call (eg proc/template etc) and instead all we get is something that can't be called (eg iterator); test cases demonstrate that.

I find behavior after this PR much friendlier; in complex cases it helps to see what went wrong